### PR TITLE
Add cache usage to sapcontrol gatherer

### PR DESCRIPTION
**Pull request by arbulu89**
Add factscache usage to sapcontrol gatherer, as sapcontrol data needs to be used in the future new gatherer to get ASCS/ERS clusters data.

It stores the entry in the cache as `sapcontrol:${method}:${sid}:${instance}`. for example: `sapcontrol:GetProcessList:PRD:00`.

This removes the need for a local cache as well.

### CODE REVIEW

File: internal/factsengine/gatherers/sapcontrol.go [hunk 1]
```diff
@@ -89,24 +91,58 @@ type SapControlInstance struct {
...
 		fs:         fs,
+		cache:      cache,
+	}
+}
+
+func (s *SapControlGatherer) SetCache(cache *factscache.FactsCache) {
+	s.cache = cache
+}
+
+func memoizeSapcontrol(args ...interface{}) (interface{}, error) {
```
Comment (arbulu89) [inline internal/factsengine/gatherers/sapcontrol.go:119]: I cannot add a test for this...
In reality, this `!ok` shouldn't never happen as this function is only called internally with known arguments.
But I just added the cases for some unexpected error, that I cannot forsee now

**Review by EMaksy — COMMENTED** [review]
Hey, hey, no go export, but the logic seems to be fully fine! 
LGTM

**Review by CDimonaco — APPROVED** [review]

File: internal/factsengine/gatherers/sapcontrol.go [hunk 2]
```diff
@@ -89,24 +91,58 @@ type SapControlInstance struct {
...
+	ctx, ok := args[0].(context.Context)
+	if !ok {
+		return nil, ImplementationError.Wrap("error using memoizeSapcontrol. Context must be 1st argument")
+	}
+
+	webService, ok := args[1].(sapcontrolapi.WebServiceConnector)
+	if !ok {
+		return nil, ImplementationError.Wrap("error using memoizeSapcontrol. WebServiceConnecter must be 2nd argument")
```
Comment (rtorrero) [inline internal/factsengine/gatherers/sapcontrol.go:127]: WebServiceConnectOr (typo)

File: internal/factsengine/gatherers/sapcontrol.go [hunk 3]
```diff
@@ -89,24 +91,58 @@ type SapControlInstance struct {
...
+	}
+
+	instanceNumber, ok := args[2].(string)
+	if !ok {
+		return nil, ImplementationError.Wrap("error using memoizeSapcontrol. string must be 3rd argument")
```
Comment (rtorrero) [inline internal/factsengine/gatherers/sapcontrol.go:132]: Capitalize after `.` (also in line 137)

**Review by rtorrero — APPROVED** [review]
LGTM

File: internal/factsengine/gatherers/sapcontrol.go (hunk 3 above)
Comment (arbulu89) [reply internal/factsengine/gatherers/sapcontrol.go:132]: This is intentional. `string` type is in small case. The others are in Capital case, as they are described this way in the code

**Pull request merged**